cas-server-core/src/main/java/org/jasig/cas/services/DefaultRegisteredServiceAuthorizationStrategy.java

```diff
@@ -0,0 +1,258 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Verify presence of service required attributes.
+     * <ul>
+     *     <li>If no required attributes are specified, authz is granted.</li>
+     *     <li>If ALL required attributes must be present, and the principal contains all and there is
+     *     at least one attribute value that matches the required, authz is granted.</li>
+     *     <li>If ALL required attributes don't have to be present, and there is at least
+     *     one principal attribute present whose value matches the required, authz is granted.</li>
+     *     <li>Otherwise, access is denied</li>
+     * </ul>
+     * Note that comparison of principal/required attributes is case-sensitive. Exact matches are required
+     * for any individual attribute value.
+     */
+    @Override
+    public boolean isServiceAccessAuthorizedForPrincipal(final Map<String, Object> principalAttributes, final Service service) {
+        if (getRequiredAttributes().isEmpty()) {
+            logger.debug("No required attributes are specified");
+            return true;
+        }
+        if (principalAttributes.isEmpty()) {
+            logger.warn("No principal attributes are found to satisfy attribute requirements for [{}]", service.getId());
+            return false;
+        }
+
+        if (principalAttributes.size() < getRequiredAttributes().size()) {
+            logger.warn("The size of the principal attributes that are [{}] does not match requirements, "
+                    + "which means the principal is not carrying enough data to grant authorization",
+                    principalAttributes);
+            return false;
+        }
+
+        logger.debug("These required attributes [{}] are examined against [{}] before service [{}] can proceed.",
+                getRequiredAttributes(), principalAttributes, service.getId());
+
+        final Sets.SetView<String> difference = Sets.intersection(getRequiredAttributes().keySet(), principalAttributes.keySet());
+        final Set<String> copy = difference.immutableCopy();
+
+        if (isRequireAllAttributes() && copy.size() < getRequiredAttributes().size()) {
+            logger.warn("Not all required attributes are available to the principal");
+            return false;
+        }
+
+        for (final String key : copy) {
+            final Set<?> requiredValues = getRequiredAttributes().get(key);
+            Set<?> availableValues = null;
+
+            final Object objVal = principalAttributes.get(key);
+            if (objVal instanceof Collection) {
+                final Collection valCol = (Collection) objVal;
+                availableValues = Sets.newHashSet(valCol.toArray());
+            } else {
+                availableValues = Collections.singleton(objVal);
+            }
+
+            final Sets.SetView<?> differenceInValues = Sets.intersection(availableValues, requiredValues);
+            if (differenceInValues.size() > 0) {
```

Comment:
The use-case that is implemented here is consistent with how the current Unicon addon functions. In fact, here we go one step further and expose an option to require all attributes. So the idea is:

(Current addon)
- Match one attribute name, whose value could match any of the required values.

or if we require all attributes:

- Check all attribute names and at least one value must match a required value.

Use cases might be the following stories:

1. To access service X, principal must have attribute Y whose value may match any of A, B or C. (memberOf=student or staff or admin)

2. To access service X, principal must have attributes Y AND Z, whose value may match any of A, B or C
(role=admin or superuser AND memberOf=admin or superuser)

What we dont have yet is:

3. To access service X, principal must have attributes Y AND Z, whose value may match A and B and C.
(role=admin AND superuser AND memberOf=admin AND superuser)

Not sure this last one is all that practical.